0027-safely-recommend-actions.adoc

```diff
@@ -0,0 +1,173 @@
+= COMIT Protocol Actions
+Tobin C. Harding <tobin@coblox.tech>;
+:toc:
+:revdate: 2020-08-07
++ Created on: {revdate} +
+
+== Introduction
+
+In order for two parties to exchange assets (swap) using the COMIT protocol each party must take certain pre-defined actions at certain times. A swap is guaranteed to be atomic if, and only if, the correct sequence of actions is followed. It is therefore paramount that a user knows what stage the swap is up to and what step they should take next. Additionally, at any stage, each party should be able to answer the following question: "If I do the next step is there time, for myself and the counterparty, to do all subsequent steps?".
+
+== Swap State Machine
+
+A successful COMIT atomic swap (happy path) may be expressed as a series of state transitions:
+
+        Created --> Alpha Funded --> Beta Funded --> Beta Redeemed --> Alpha Redeemed --> Done
+
+Each transition is associated with a specific action named accordingly e.g., 'fund alpha' transitions from created to alpha funded.
+
+For each state transition there is a time requirement, henceforth called the transition period. The total time requirement for a swap is the sum of the transition periods. A transition period is effected by the state we are transitioning from and the state we are transitioning to, each transition period therefore is calculated differently.
+
+== Transition Period
+
+There exist a number of factors, that when combined, represent the transition period.
+
+.Considerations
+* Time to Act (`TTA`)
+* Time to Inclusion (`TTI`)
+* Time to Final (`TTF`)
+* HTLC Expiry time
+
+=== Time to Act
+
+For each actor there is a time frame within which they can be expected to act.
+
+Different users will have different response times e.g., a long running daemon is likely to respond to swap transitions more quickly than a user who is manually checking an application before doing each action. We define Time To Act (`TTA`) as the window of time within which the counterparty can expected to take the next action.
+
+=== Time To Inclusion
+
+For each blockchain there exists a time for a transaction to be included in a block.
+
+The likelihood of a transaction being included in a block is dependant on the transaction fee coupled with the current load on the blockchain network i.e., the number of current unconfirmed transactions. An increase in the transaction fee leads to an increase in the probability that the transaction will be included. Inclusion is not guaranteed, therefore the probability is defined to be the percentage chance of a transaction being included in a block within the next `N` blocks.
+
+With a given fee and a given probability tolerance we define Time To Inclusion for a given blockchain as:
+
+----
+TTI = N * <average block time>
```

Comment:
We should mention that N depends on the amount, i.e. I would ask for at least 6 confirmations for $1mio transaction vs. 1 confirmation is fine for a $1000 transaction.

Reply:
I don't think N depends on a concrete amount. It depends on the perceived value of the amount. I may think that $1000 is a lot of money and needs 20 confirmations on Bitcoin but you may consider $1000 to be pocket change and be happy with 1 confirmation. For me this is implied in the configuration of number of required confirmations.

Is anyone even going to read this spike again after the implementation is done? I doubt it, since it does not live with the code (as I originally hoped) it is just an artefact of the dev process to me. I can definitely add a section on further configuration considerations that folks may wish to use.

Reply:
> I don't think N depends on a concrete amount. It depends on the perceived value of the amount. I may think that $1000 is a lot of money and needs 20 confirmations on Bitcoin but you may consider $1000 to be pocket change and be happy with 1 confirmation. For me this is implied in the configuration of number of required confirmations.

The reason why the amounts matter is because _attacking_ a blockchain costs money. A 1h/6 blockchain reorg for Bitcoin currently costs ~$500k in hashing power which means if you have a $1mio tx it's worth to invest those $500k into a 51% attack and create a 6 block deep re-org to double spend your $1mio transaction. On Ethereum a 1h attack on the otherhand would only cost you about $300k. 
It's not worth to spend $500k on a reorg if you want to double spend a $1000 tx. 
This means, the finality for a transaction of a blockchain depends on the amount and the current hash rate cost. 

A nice table on cost: https://www.crypto51.app/
A recent example of a 51% attack: https://www.coindesk.com/ethereum-classic-suffers-second-51-attack-in-a-week

Reply:
Cool, I did not know about this. I don't really think that we need to consider transactions with value greater than $300k at the moment when we do not even have a mainnet solution. I would think that it should be noted somewhere in the docs though. Thanks for educating me. Would documentation of this consideration be enough for you?